module/zfs/dsl_deadlist.c

```diff
@@ -957,23 +963,39 @@ dsl_livelist_iterate(void *arg, const blkptr_t *bp, boolean_t bp_freed,
 
 	if ((t != NULL) && (zthr_has_waiters(t) || zthr_iscancelled(t)))
 		return (SET_ERROR(EINTR));
+
+	livelist_entry_t node;
+	node.le_bp = bp;
+	livelist_entry_t *found = avl_find(avl, &node, NULL);
 	if (bp_freed) {
-		livelist_entry_t *node = kmem_alloc(sizeof (livelist_entry_t),
-		    KM_SLEEP);
-		blkptr_t *temp_bp = kmem_alloc(sizeof (blkptr_t), KM_SLEEP);
-		*temp_bp = *bp;
-		node->le_bp = temp_bp;
-		avl_add(avl, node);
-	} else {
-		livelist_entry_t node;
-		node.le_bp = bp;
-		livelist_entry_t *found = avl_find(avl, &node, NULL);
-		if (found != NULL) {
-			avl_remove(avl, found);
-			kmem_free((blkptr_t *)found->le_bp, sizeof (blkptr_t));
-			kmem_free(found, sizeof (livelist_entry_t));
+		if (found == NULL) {
+			/* first free entry for this blkptr */
+			livelist_entry_t *e =
+			    kmem_alloc(sizeof (livelist_entry_t), KM_SLEEP);
+			blkptr_t *temp_bp =
+			    kmem_alloc(sizeof (blkptr_t), KM_SLEEP);
+			*temp_bp = *bp;
+			e->le_bp = temp_bp;
```

Comment:
I didn't see a good reason for `le_bp` to be allocated separately from `e` (as opposed to embedded within it).  This is orthogonal to your change, but if you wanted to change that while you're here, it would have the nice property of actually *reducing* the total amount of memory required for the AVL tree, despite adding the 4-byte refcount.

Reply:
Great point! Done